File: packages/desktop/Sources/AutopodClient/Types/ExecutionProvenance.swift
import Foundation

public struct ExecutionProvenanceResponse: Codable, Sendable { public let latest: ExecutionProvenance? }
public struct ExecutionProvenance: Codable, Sendable {
  public let executionId: String
  public let generation: Int
  public let checkedAt: String
  public let purpose: String?
  public let subject: String?
  public var subjectLabel: String { subject == "reviewer" ? "Reviewer" : "Configured worker" }
  public let status: String
  public let version: Int?
  public let surface: String?
  public let dispatchModel: String?
  public var runtimeLabel: String { surface == "provider-api" ? "Provider API · dispatch model \(dispatchModel ?? "unverified")" : "\(runtime ?? "unverified") CLI \(cliVersion ?? "unverified")" }
  public var imageLabel: String { surface == "provider-api" ? "not applicable" : (imageDigest ?? "unverified") }
  public let runtime: String?
  public let model: String
  public let providerId: String?
  public let providerAccountId: String?
  public let cliVersion: String?
  public let imageDigest: String?
  public let contractHash: String
  public let validationImplementationHash: String?
  public let release: DaemonReleaseSnapshot
  public let capabilities: ExecutionCapabilities
  public let commands: ExecutionCommandPreflight
  public let diagnostics: [ExecutionDiagnostic]
}
public struct ExecutionCapabilities: Codable, Sendable {
  public let streamingExec: String
  public let memoryLimitBytes: Int?
  public let cpuLimit: Double?
  public let networkMode: String?
}
public struct ExecutionCommandPreflight: Codable, Sendable {
  public let requirements: [ExecutionCommandRequirement]
  public let unresolvedSources: [String]
  public let deferredArtifacts: [String]
  public let explicitDependencies: Bool
}
public struct ExecutionCommandRequirement: Codable, Sendable {
  public let source: String
  public let executable: String
  public let available: Bool?
}
public struct ExecutionDiagnostic: Codable, Sendable {
  public let code: String
  public let detail: String
}

File: packages/shared/src/types/execution-provenance.ts
import type { DaemonReleaseInfo } from './release.js';
import type { RuntimeType } from './runtime.js';

export interface ExecutionProvenance {
  id: string;
  version: 1 | 2;
  /** v2 distinguishes direct API and host CLI reviewer dispatch preparation. */
  surface?: 'provider-api' | 'host-cli';
  dispatchModel?: string;
  podId: string;
  taskId: string;
  executionId: string;
  generation: number;
  checkedAt: string;
  status: 'checked' | 'blocked';
  /** Absent on early v1 records; those describe coding startup. */
  purpose?: 'coding' | 'validation' | 'review' | 'completion';
  /** Absent on historical records, which probed the configured worker even for review. */
  subject?: 'worker' | 'reviewer';
  runtime: RuntimeType | null;
  model: string;
  providerId: string | null;
  providerAccountId: string | null;
  release: DaemonReleaseInfo;
  cliPath: string | null;
  cliVersion: string | null;
  imageDigest: string | null;
  contractHash: string;
  validationImplementationHash: string | null;
  capabilities: {
    streamingExec: 'supported' | 'unsupported' | 'unverified';
    memoryLimitBytes: number | null;
    cpuLimit: number | null;
    networkMode: string | null;
  };
  commands: {
    requirements: Array<{ source: string; executable: string; available: boolean | null }>;
    unresolvedSources: string[];
    deferredArtifacts: string[];
    explicitDependencies: boolean;
  };
  diagnostics: Array<{ code: string; detail: string }>;
}
export type ExecutionProvenanceInput = Omit<
  ExecutionProvenance,
  'id' | 'podId' | 'taskId' | 'executionId' | 'generation' | 'checkedAt'
>;
